### PR TITLE
Make fixture initializer wait for gateway being registered

### PR DESCRIPTION
**Pull request by elsirion**
This is my best guess at why `lightning_gateway_claims_refund_for_internal_invoice` fails sometimes. https://github.com/fedimint/fedimint/issues/222#issuecomment-1385766750

### CODE REVIEW

File: integrationtests/tests/fixtures/mod.rs [hunk 1]
```diff
@@ -248,6 +248,16 @@ pub async fn fixtures(num_peers: u16) -> anyhow::Result<Fixtures> {
...
             .await;
 
+            // TODO: deduplicate by moving fixture constructor out of match
+            // Wait till the gateway has registered itself
+            loop {
+                let gateways = user.client.fetch_registered_gateways().await?;
+                if !gateways.is_empty() {
+                    break;
+                }
+                tokio::time::sleep(Duration::from_millis(100)).await;
```
[the remark this answers is not in the document]
Comment (elsirion) [reply integrationtests/tests/fixtures/mod.rs]: Added :)

**Comment by elsirion** [comment]
Just saw a failure again … trying to reproduce

**Review by dpc — DISMISSED** [review]

**Comment by elsirion** [comment]
So apparently The output of `fetch_registered_gateways` could become empty after it had entries previously, so I now call the function that will save the GW in the DB, so that later `fetch_registered_gateways` won't be called again.

**Comment by dpc** [comment]
> So apparently The output of fetch_registered_gateways could become empty after it had entries previously, so I now call the function that will save the GW in the DB, so that later fetch_registered_gateways won't be called again.

@okjodom I thought this is not possible?

Could it be because the mint restarted or something?

**Comment by okjodom** [comment]
> > So apparently The output of fetch_registered_gateways could become empty after it had entries previously, so I now call the function that will save the GW in the DB, so that later fetch_registered_gateways won't be called again.
> 
> @okjodom I thought this is not possible? I've seen the same thing happening when I was debugging it, but I was told we have not even a route to deregister.
> 
> Could it be because the mint restarted or something?

Makes sense if mint restart cleared previous state. Gateway registration with the federation would be cleared so this I'd expect a need to re-register the gateway

**Comment by dpc** [comment]
> Makes sense if mint restart cleared previous state. 

We should persist it maybe?

**Comment by okjodom** [comment]
> > Makes sense if mint restart cleared previous state.
> 
> We should persist it maybe?

Or just make sure to set up the fixtures properly again, including gateway re-registration. Do we have a new test scenario that restarts the mint?

**Comment by elsirion** [comment]
> @okjodom I thought this is not possible? I've seen the same thing happening when I was debugging it, but I was told we have not even a route to deregister.

One more possibility: only one guardian has processed the GW registration, so it's part of the union request that returned the union of t of n guardians' responses. The next time this one guardian doesn't respond and we get the error.

**Comment by dpc** [comment]
Another possibility is that my ifederationapi broke something about the union strategy. But I think I've seen these failures sporadically before that change.

**Comment by dpc** [comment]
> Or just make sure to set up the fixtures properly again, including gateway re-registration. Do we have a new test scenario that restarts the mint?

I thought that each test starts its own fediming and gateways.

**Review by okjodom — APPROVED** [review]

**Pull request merged**